### ISSUE TITLE
Codegen for some bitflag checks could be optimized

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/coreclr#27712]
More info: https://github.com/dotnet/coreclr/pull/27603#discussion_r341567417

Sample code:

```cs
public static bool HasHighBitSet(uint value) {
    return (value & 0x8000_0000) != 0;
}
```

Actual codegen:

```asm
    L0000: test ecx, 0x80000000
    L0006: setnz al
    L0009: movzx eax, al
    L000c: ret
```

Replacing the first two instructions with `test ecx, ecx` and `setl al` would save 4 bytes of codegen. It's a minor savings but might help if this pattern occurs within a hot loop.

**Comment by mikedn** [comment]
> Replacing the first two instructions with test ecx, ecx and setl al would save 4 bytes of codegen. 

Actually this specific case should generate:
```asm
mov eax, ecx
shr eax, 31
ret
```
The test approach is only useful if such code is used as branch condition, then it would generate `test` and `js`/`jns`.

**Comment by omariom** [comment]
> if such code is used as branch condition

The codegen of the inlined method [looks good ](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGPKQAIyGBhBgbxoe6foYEsAdhgYBZABQBXISIBu2ADaSYASgZcenajx0CAZg3EAJPEf4BzABYAhfhgDKMDOPlLVKjbt3EA7AwAMCADMQQDcnt5+gT4+4do8AL4REXSMwBAQCgwmuGZWtg5OUjIMrspqWjq+hmUwDABkAQgAHP5tAPptbWoAhAC8AXFJ1AlAA===)

**Comment by mikedn** [comment]
> The codegen of the inlined method looks good 

No, the point is that this could generate `test reg, reg` instead of `test reg, imm` to avoid the 32 bit immediate.

**Comment by GrabYourPitchforks** [comment]
> The test approach is only useful if such code is used as branch condition

Yeah, this suggestion was for branch conditions specifically. I thought there was already a tracking issue for more general branchless bit manipulation of Boolean conditions?

**Comment by mikedn** [comment]
There definitely are other related issues - #7641 and #8462 come to mind and there may be others. I also have some code written that does some of this, gotta check what state it's in.
